lib/phabricator/phab_object.rb

```diff
@@ -2,18 +2,126 @@
 
 module Phabricator
   class PhabObject
-    attr_reader :phid
+    def self.inherited(other)
+      other.instance_eval do
+        @props = {}
+        @translatable_prop_names = []
+
+        prop :phid
+        attr_accessor :attrs
+      end
+    end
+
+    def self.prop(prop_sym, opts={})
+      @props[prop_sym] = opts
+      # For now, make props read-only, since we don't provide a method to save a changed object
+      define_method(prop_sym) {@attrs[prop_sym]}
+
+      name_prop = opts[:name_prop]
+      if name_prop
+        klass = opts[:class]
+        unless klass
+          raise ArgumentError.new("class is required when name_prop is provided")
+        end
+        @translatable_prop_names << prop_sym
+        define_method(name_prop) do
+          val = send(prop_sym)
+          if val.is_a?(Array)
+            val.map {|v| klass.name_from_value_value(v)}
```

Comment:
Am I reading this correctly that it will only work with values from an array and then strings, since it will stringify (like for tasks?)

What about hashes, and other non-string values?